### PR TITLE
Fixes to the file.source_list when checking for existence of HTTP source files

### LINKED ISSUE #58881: [BUG] 3002.1 file.manage UnicodeDecodeError
**Description**
In salt 3002.1 when downloading zip file there is **UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf9 in position 169: invalid start byte** I would like to point that state that downloads zookeper (tar.gz) works fine
i downgraded salt-minion version to 3001.3 and all works fine

```
----------
          ID: download_zookeeper
    Function: file.managed
        Name: /opt/apache-zookeeper-3.5.6-bin.tar.gz
      Result: True
     Comment: File /opt/apache-zookeeper-3.5.6-bin.tar.gz updated
     Started: 13:57:32.133160
    Duration: 721.194 ms
     Changes:   
              ----------
              diff:
                  New file
              mode:
                  0644

          ID: download_storm
    Function: file.managed
        Name: /opt/apache-storm-2.1.0.zip
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python3/dist-packages/salt/state.py", line 2154, in call
                  *cdata["args"], **cdata["kwargs"]
                File "/usr/lib/python3/dist-packages/salt/loader.py", line 2188, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python3/dist-packages/salt/states/file.py", line 3185, in managed
                  source, source_hash = __salt__["file.source_list"](source, source_hash, __env__)
                File "/usr/lib/python3/dist-packages/salt/modules/file.py", line 4376, in source_list
                  query_res = salt.utils.http.query(single, method="HEAD")
                File "/usr/lib/python3/dist-packages/salt/utils/http.py", line 656, in query
                  result_text = result_text.decode("utf-8")
              UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf9 in position 169: invalid start byte
     Started: 13:57:32.856475
    Duration: 6269.537 ms
     Changes:   
```

**Setup**
SLS:
```
download_zookeeper:
  file.managed:
    - name: /opt/apache-zookeeper-{{ salt['pillar.get']('packages:zookeeper_version', '3.4.9') }}-bin.tar.gz
    - source:
      - http://archive.apache.org/dist/zookeeper/zookeeper-{{ salt['pillar.get']('packages:zookeeper_version', '3.4.9') }}/apache-zookeeper-{{ salt['pillar.get']('packages:zookeeper_version', '3.4.9') }}-bin.tar.gz
    - source_hash: cc6a1d11129e8330141b756625bfdba0
    - onfail:
      - id: chcek_download_zookeeper

chcek_download_storm:
  file.exists:
    - name: /opt/apache-storm-{{ salt['pillar.get']('packages:storm_version', '2.1.0') }}.zip

download_storm:
  file.managed:
    - name: /opt/apache-storm-{{ salt['pillar.get']('packages:storm_version', '2.1.0') }}.zip
    - source:
      - https://www-eu.apache.org/dist/storm/apache-storm-{{ salt['pillar.get']('packages:storm_version', '2.1.0') }}/apache-storm-{{ salt['pillar.get']('packages:storm_version', '2.1.0') }}.zip
    - source_hash: ae3047bb130aa92dda900b2d851d1cdd
    - onfail:
      - id: chcek_download_storm
```
Pillar:
```
packages:
  zookeeper_version: 3.5.6
  storm_version: 2.1.0
```
**Steps to Reproduce the behavior**
salt-call state.apply

**Expected behavior**
Both files are downloaded and created. Output from version 3001.3:
```
----------
          ID: download_zookeeper
    Function: file.managed
        Name: /opt/apache-zookeeper-3.5.6-bin.tar.gz
      Result: True
     Comment: File /opt/apache-zookeeper-3.5.6-bin.tar.gz updated
     Started: 14:05:30.054014
    Duration: 327.184 ms
     Changes:   
              ----------
              diff:
                  New file
              mode:
                  0644
----------
          ID: chcek_download_storm
    Function: file.exists
        Name: /opt/apache-storm-2.1.0.zip
      Result: False
     Comment: Specified path /opt/apache-storm-2.1.0.zip does not exist
     Started: 14:05:30.381339
    Duration: 1.242 ms
     Changes:   
----------
          ID: download_storm
    Function: file.managed
        Name: /opt/apache-storm-2.1.0.zip
      Result: True
     Comment: File /opt/apache-storm-2.1.0.zip updated
     Started: 14:05:30.382959
    Duration: 1509.718 ms
     Changes:   
              ----------
              diff:
                  New file
              mode:
                  0644
```

**Versions Report**
```
salt-call --versions
Salt Version:
           Salt: 3002.1
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: 1.3.10
      pycparser: Not Installed
       pycrypto: Not Installed
   pycryptodome: 3.4.7
         pygit2: Not Installed
         Python: 3.6.9 (default, Oct  8 2020, 12:12:24)
   python-gnupg: 0.4.1
         PyYAML: 3.12
          PyZMQ: 17.1.2
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5
 
System Versions:
           dist: ubuntu 18.04 Bionic Beaver
         locale: UTF-8
        machine: x86_64
        release: 4.15.0-122-generic
         system: Linux
        version: Ubuntu 18.04 Bionic Beaver
```
</details>

**Additional context**

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
When calling salt.utils.http.query with a HEAD method to check for the existence of a source, default to using the requests library as the tornado library switches the method from HEAD to GET if the original URL is a 3xx redirect.

### What issues does this PR fix or reference?
Fixes: #58881

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/modules/file.py [hunk 1]
```diff
@@ -4315,7 +4321,21 @@ def source_list(source, source_hash, saltenv):
                         ret = (single_src, single_hash)
                         break
                 elif proto.startswith("http") or proto == "ftp":
-                    query_res = salt.utils.http.query(single_src, method="HEAD")
+                    # If rquests is available we'll use that for the HEAD request in case the URL ends up being a
```
Comment (Ch3LL) [inline salt/modules/file.py:4324]: What about for users that do not have requests installed by default? Will they still continue to get the traceback in the issue this PR resolves? Is there a way to catch this 3xx redirect is occurring and return an error letting the user know they should use requests?

Comment (dequis) [reply salt/modules/file.py:4324]: Yeah, the `decode_body=False` part of the `else` clause fixes it when not using requests. I just patched my local salt and it worked with just

```diff
-query_res = salt.utils.http.query(single_src, method="HEAD")
+query_res = salt.utils.http.query(single_src, method="HEAD", decode_body=False)
```

But I guess the rest of the PR doesn't hurt? Other than the part where throwing more details in delays its own merging.

Comment (Ch3LL) [reply salt/modules/file.py:4324]: thanks for testing that @dequis . So it seems we do not necessarily need requests, but just need to pass in `decode_body=False`. Does that align with your findings @garethgreenaway ?

Comment (Ch3LL) [reply salt/modules/file.py:4324]: bump @garethgreenaway ^

Comment (garethgreenaway) [reply salt/modules/file.py:4324]: Apologies for the delay on this one.  I was able to verify that including `decode_body=False)` in the call to salt.utils.http.query fixes it with both `requests` and `torando`.  I've updated the PR to include the changes.

File: salt/modules/file.py [hunk 2]
```diff
@@ -4315,7 +4315,9 @@ def source_list(source, source_hash, saltenv):
                         ret = (single_src, single_hash)
                         break
                 elif proto.startswith("http") or proto == "ftp":
-                    query_res = salt.utils.http.query(single_src, method="HEAD")
+                    query_res = salt.utils.http.query(
+                        single_src, method="HEAD", decode_body=False
```
Comment (Ch3LL) [inline salt/modules/file.py:4319]: Thinking this one a bit more thoroughly, I imagine there could be a source that would require the body to be decoded?

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/file.py (hunk 2 above)
Comment (garethgreenaway) [reply salt/modules/file.py:4319]: This is simply checking that the file exists on the remote end there and it's a valid source, there shouldn't be any resource we need to decode the body.

**Review by Ch3LL — DISMISSED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**